### ISSUE TITLE
"FCModel 2" non-unique-instances branch

### ISSUE THREAD

**Issue by marcoarment** [marcoarment/FCModel#98]
Just pushed the first commit e0f9b3d3 of what might be called "FCModel 2" to the "non-unique-instances" branch. This is a **big, likely-breaking change** that applies a lot of lessons learned since the original design:

**Instances are no longer tracked or kept unique in memory. All instances are "detached".**
- You can have two instances in memory of the same logical model (same table and primary key).
- Changes made to a model object or table won't affect any other copies in memory.
- If you want your interface to react to changes, you must reload any data in response to FCModel's change notifications.
- `dataWasUpdatedExternally`, reloads, and conflict resolution are no longer necessary and have been removed.
- `allLoadedInstances` is no longer possible and has been removed.
- Transactions are now supported, although limited. (Rollbacks don't revert changed models in memory.)
- Autoincrement, formerly deprecated, is now removed and will raise an exception.

**All saves and deletes are expected to succeed. Saves shouldn't unexpectedly fail or be blocked.**
- Removed `should/didInsert/Update/Delete`, `saveWasRefused`, `saveDidFail`. To customize behavior, override `save` and call `[super save]` from the subclass.
- `FCModelSaveResult` is removed. `delete` now returns void and `save` now returns a BOOL to indicate whether changes were made.
- Saves and deletes that fail from database errors now raise exceptions. `lastSQLiteError` is removed.
- All FCModel exceptions now have name `FCModelException`.
- `saveAll` has been removed. Saves should happen intentionally, right after you change the data.
- Custom serializations for `NSURL`, `NSDate`, etc. have been removed to avoid a lot of subtle bugs and inconsistencies between formats.

**Notifications have been simplified to just a single FCModelChangeNotification for any change to a table.**
- `FCModelInstanceSetKey` has been replaced by `FCModelInstanceKey` and will always be one instance if known, or unset for many/unknown.
- Notification-batching functions are now redundant and have been removed since transactions also batch notifications.

Take a look and see what you think so far. What other changes should we consider now that we're breaking things?

**PLEASE don't ship any apps with this yet.**

**Comment by mirion** [comment]
I think that the framework should be able to offer support for easy values transformation. As @lickel mentioned above, this is a highly repetitive task. Indeed not having it is not terrible but annoying and for me it is close to bad.

This is related to another aspect: for the time being the data model is created through inspection of the database table and class structures. This is not always very confortable. It would be nice to declare a sort of descriptor for a given pair class/table that is able to map attributes to columns. This descriptor may also declare the transformers. But before this descriptor, restoring the ability to transform data is (in my opinion) something really necessary.

**Comment by Ahti** [comment]
I agree with @mirion that value transformations would be nice to have.

Even though I removed my one implementation of `-unserializedDatabaseRepresentationOfValue:` (in one of ten model classes) recently (see #104), I still rely on the current implementation handling NSDate for me, and think writing all the boilerplate to handle transforming properly would be annoying and error-prone.

I think an approach similar to [how CoreData handles this](https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/CoreData/Articles/cdNSAttributes.html) could turn out nicely. Basically it allows you to specify a NSValueTransformer, and by default uses a transformer that knows how to handle objects implementing `NSCoding`.

The `NSCoding` part would cover all the Classes the old implementation handled, plus a big bunch of others.

**Comment by andreyvit** [comment]
I just went to open a ticket asking about typical expected FCModel concurrency patterns, and found this.

My two cents:

0. Loving FCModel, been using it a lot since last summer. (Also loving Overcast and atp.)

1. Detached instances are a great idea, because otherwise there's no convenient way to perform background operations involving DB instance modifications while the UI keeps reading them. I actually wonder how Overcast handled this before. Did you make all your properties atomic and rely on no fatal inconsistencies arising along the way?

2. I love built-in serialization. I've actually used those overridable methods to customize serialization formats in the past, to support my own calendar-related class that acted like a value. NSDates serialized as numbers can probably always be parsed back, no nil issues there.

    I disagree that wrapper properties result in a simpler approach overall, compared to serialization and deserialization methods. Like @mirion said, it's a highly repetitive task that is worth automating.

    I also disagree that removing serialization is a correct approach conceptually. I _am_ storing a value whose domain is e.g. “anything representable by NSURL”. The fact that I can't make some sort of a check constraint on SQLite side to enforce the domain doesn't mean that I cannot rely on the fact that the values are correct; I'm just pushing the enforcement to the application layer, and FCModel should make sure that anything that can be serialized can also be deserialized.

    If someone bypasses FCModel serialization and puts an incorrect value into their database, they've destroyed the integrity of their data and should enjoy a crash when reading it back.

3. Releasing as v2.0 of the pod makes perfect sense, and is the only right approach. Major version changes are supposed to introduce breaking changes.

I wonder if Overcast is already shipping FCModel 2.0? Any particular reason why we shouldn't ship our apps with it as well, provided that we're prepared to debug the problems?

**Comment by andreyvit** [comment]
Oh, one more thing. I'm using those keyed access methods a lot; am I the only one? Of course, this is a minor issue (I can use my collection grouping categories, or I can add a category to FCModel), but I wonder why @marcoarment finds them to bring “too little gain”.

My use case is bulk-fetching of data (often on a background queue) before processing it (on the main queue).

**Comment by mirion** [comment]
Some more comments:
1. I missed the removal of instancesFromResultSet. Since FCModel is not providing any method to execute a random query and extract the results into an array, I'm finding this decision very inappropriate. Also the argument that it simplifies the imports is superfluous because FCModel already needs DMDatabase.h

2. Also, another unfortunate decision, why raise exceptions? This kind of error management is completely useless for those using this library in Swift. And beyond Swift, raising exceptions is a little bit "oldie". Why change the errors management when the library that FCModel wraps is using error codes?

FCModel is a highly useful library, please don't dumb it down. Of course, we can fork, but it is not the same.
Thanks

**Comment by marcoarment** [comment]
Sorry, been traveling a lot and haven't had time to respond. Some quick notes, though:

@mirion 1: The general idea here was to ensure that every column is always selected (`SELECT *`), with no joins or aliases or anything that could mess up the decoding of the result set into saturated model instances. This is less important to core functionality since the move to non-unique instances, but it's still important for caching opportunities and avoiding weird bugs.

@mirion 2: We're using exceptions idiomatically like Cocoa: they're intended to signify errors during development, are not intended to be caught, and are expected not to occur in normal operation of the app. Error codes, NSError out params, etc. are for more common, expected, and/or recoverable errors.

@andreyvit: My idea for removing serialization has not been well-received, and some counterarguments are strong. I'm considering the use of NSValueTransformer to handle custom serialization needs, with FCModel likely providing default implementations for NSURL and NSDate like before, but in a better-structured way.

Overcast has been shipping with this branch since version 1.1.2 a few months ago. I had some concurrency bugs initially, but the move to using `reloadAndSave` for all writes fixed them all and has been rock-solid. It's so good that I'll likely remove external access to `save` entirely, forcing all writes to go through `reloadAndSave`, since it's so much better and eliminates a lot of potential bugs.

I'd love for all model instances to be immutable outside a `reloadAndSave` context, but I'm not sure if there's a great way to do that without clunky hacks or custom code in each model subclass. Ideally, the properties would all be declared `readonly` so Xcode and the compiler would enforce it, but that makes writes potentially hacky. Suggestions welcome.

I'm also considering bringing back one of the benefits of unique instances in a way, but optionally. We'd have something like a `persistentCopy` method on each model instance that would keep itself updated and have its properties be KVO-observable.

So before this is merged and declared 2.0, I'd like to experiment with:

* NSValueTransformer for value serialization
* Immutable-model possibilities
* `persistentCopy`

...which I'll start doing shortly. Again, suggestions welcome.

**Comment by Ahti** [comment]
Regarding NSValueTransformer things: CoreData uses, by default, a reversed version of `[NSValueTransformer valueTransformerForName:NSKeyedUnarchiveFromDataTransformerName]`, so that would probably be all that's needed for a default implementation by FCModel.

**Comment by johncblandii** [comment]
How's this looking @marcoarment? I've been riding on "1.0" but am seeing some bad performance for larger data sets (really not that big, less than 100 objects) and hoping this could help.

Any updates?

**Comment by johncblandii** [comment]
Dude, integrated non-unique and the performance went from taking 40 seconds to save everything to 1.4s. Yummy!

**Comment by mirion** [comment]
In fact, your problem is triggered by the huge number of notifications that are generated by FCModel. Just modify the code to disable the notifications. II already have an implementation, if you need it, let me know and I will make a pull request.

**Comment by johncblandii** [comment]
The timing issues were because of `dataWasUpdatedExternally`. The biggest FCModel time suck now is around notifications but it is sub-500ms so I'm not sweating it too much right now.

**Comment by mirion** [comment]
@marcoarment are you still maintaining FCModel 1? If yes, I have a several corrections into my fork. If you want to integrate something, please let me know to make PRs.

**Comment by johncblandii** [comment]
No, I'm on non-unique now and not turning back! :-D

**Comment by mirion** [comment]
@johncblandii Indeed everything starts with dataWasUpdatedExternally, but dig deeper, you will see that the problem is eventually produced by the notifications. I profiled the code.

In my opinion, (at least for my use case), v2 has some big flaws and is not ready for prime time. Your experience my be different.

**Comment by johncblandii** [comment]
It isn't a full release so I expect some issues (like the dropping of serialization which I'm addressing now) but overall the performance is worth it. Seeing as Overcast is using it in a large way, I'm confident I can make it work for this app.

**Comment by marcoarment** [comment]
@mirion: I don't plan any more updates to v1 — v2 is the way forward. I'd love to hear the flaws you've identified in it.

**Comment by mirion** [comment]
@marcoarment 
I already mentioned my problems:
1. the removal of instancesFromResultSet. This is useful for complex queries, for example for FTS. I think that removing functionality in order to protect the developers is not really the right path.
2. the use of exceptions on save/delete. For me this is a very big issue. You may get runtime errors for example related to PKs or other constraints that are not predictable during the development phase. For example after a sync into a distributed environment, etc.
3. the serialization issue

There were others, but I can't remember right now. Will write again if I'll remember.
In my case, all these issues are blocking. It looks that for others the implications are much smaller.

Regarding v1, I have some corrections/improvements that may be useful for v2:
https://github.com/mirion/FCModel/tree/mirion_dev/fix_inExpectedWrite
https://github.com/mirion/FCModel/tree/mirion_dev/write_transaction_on_queue
and also views support:
https://github.com/mirion/FCModel/tree/mirion_dev/views_file_support

**Comment by johncblandii** [comment]
I'm curious @mirion, are you trying to use a remote value as a PK?

**Comment by marcoarment** [comment]
Sounds like we just disagree on the meaning of "has some big flaws and is not ready for prime time". I took that to mean _bugs_, while you appear to be referring to features and implementation details on which we disagree.

**Comment by lynns** [comment]
@marcoarment Do you have a sense for when the non-unique branch will be ready for production use?  We're anxious to switch over but have been holding off for the final changes to be done before we do.  Is the branch pretty stable as is or should we keep holding off for a while longer?

Thanks for all the work you've done on this, we've really enjoyed using FCModel in our apps.

**Comment by marcoarment** [comment]
@lynns I haven't had time to formally publish and document it as such, but I consider this branch stable for production. I've been shipping it in Overcast for months with no problems.

**Comment by lynns** [comment]
@marcoarment Sounds good to me.  We'll start trying it out and see how it goes for us.  Thanks!

**Comment by johncblandii** [comment]
@lynns I ported to non-unique last night and I can speak highly enough about it. You'll need to address serialization (I just used `setValue:forKey:` to deal w/ one date issue) and that was pretty much it.

It's well worth at least a spike.

@marcoarment, I think the community could help document it. Hopefully these projects wind down in the next week or so and I can lend a hand in that area.

**Comment by mirion** [comment]
Please don't understand me wrongly. My words may sound too strong, I really think that you did a great job with this library. I like it, I'm using it and thanks a lot for all your work. If I'm commenting your decisions is not to criticize you, but just because I like your code, I want to use it in the future, therefore I want to see it in the best shape for as many of us (including for advanced use), not just for some cases.

I used the word "flaw" because some design decisions (every api that was removed represent in the end a design decision) are beyond the term "bugs". Of course, these design decisions are in the end easy to revert and solve the bugs ;-)

@johncblandii I had a problem (in v1) with records sent from the server, where because of some networking problems, the server was pushing them two times. Just a pipe problem, completely unpredictable.

**Comment by prendio2** [comment]
`FCModelInstanceKey` is documented but seems to be unimplemented as of yet. I'm happy to look into addressing this but wanted to check and see if this was an oversight or if you ran into any complications with it? I can open this as a new issue if you prefer.

**Comment by marcoarment** [comment]
@prendio2 Ah, that's legacy. The notifications used to track which instances changed and would include them in the notification payloads, but they don't anymore.

**Comment by prendio2** [comment]
@marcoarment did they ever do that on this non-uniques-instances branch? The initial commit https://github.com/marcoarment/FCModel/commit/e0f9b3d3b58f132f43171d3defcf18e66b7c2af3 doesn't use them in the userInfo dictionary either.

Would you be open to having them in the main repo if I (re)implement it and pull request or is that something you'd prefer I kept in my own fork?

**Comment by marcoarment** [comment]
@prendio2 No, this branch never included them. I'd like to keep them out — the rationale was not only to improve performance on changes and reduce memory overhead, but to reinforce the non-unique-instances design pattern, in which you're intended to always manage your own instances on the caller side and never assume any one instance is shared (ever) or up-to-date (outside of a reloadAndSave block).

**Comment by johncblandii** [comment]
@mirion, I store all remote `id` fields as `remote_id` for all models and find/reference that field throughout the app, even for "joins". For hundreds of objects, I see no collisions. Unsure if that's related but it's helped quite a bit on two separate apps.

**Comment by prendio2** [comment]
@marcoarment OK. My rationale is I want to get the id of the changed object to avoid doing a complete reload each time any object changes. Maybe I'll add an id to the user info dict rather than the FCModel object itself.

**Comment by mirion** [comment]
@johncblandii That was just an example of error that is effectively unpredictable. If the app crashes because the code raises an exception, this doesn't give any good impression. Such an error should be recoverable. In my case I just changed the table definition to use on conflict replace. Problem solved, even without error handling (via return nserror) but we can't conceive all these edge cases that are completely out of our control.

**Comment by lynns** [comment]
@marcoarment You mentioned previously that you are considering removing external access to `save` in favor of always using `reloadAndSave`.  That makes sense in update cases but I'm curious how you handle object creation where you want to create a new object, set properties on it, and then save it to the db.  `reloadAndSave` doesn't make a lot of sense in that case.  Thoughts?

**Comment by marcoarment** [comment]
After banging my head against the wall trying to solve some obscure and tricky Overcast bugs related to non-unique-instances for months, I've decided to try a different approach in the new **unique2** branch: it's effectively all of the other improvements from **non-unique-instances**, but with two big changes:

* Unique instances again.
* The DB queue is now the main thread.

Separately, these are weird and can be problematic. Together, I think they make a lot more sense.

Unique instances allow **far** simpler, faster, and more efficient code for many tasks, but hit weird issues in FCModel 1 with the instance's field values changing from another thread in the middle of an operation. The recent introduction of `reloadAndSave` went a long way toward fixing this. And by serializing all DB access onto the main thread, all operations on the main thread, which is the vast majority or entirety of all DB operations in most apps, **can't** have this happen to them.

Maybe there's something I'm missing here, but I'm trying this in Overcast betas this summer and will report back.

**Comment by lynns** [comment]
@marcoarment Any updates on how your `unique2` branch is working?

**Comment by marcoarment** [comment]
@lynns: I'm using it exclusively now, and about to ship Overcast 2.0 with it. So far, it's rock-solid — better than 1.0 _and_ non-unique-instances.

**Comment by Ahti** [comment]
Now that you (and maybe others?) have been using this for a while, would you consider releasing this as a FCModel 2.0?
